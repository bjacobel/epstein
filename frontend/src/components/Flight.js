import React from 'react';
import { useQuery, gql } from '@apollo/client';

const FLIGHT = gql`
  query Flight($id: Int!) {
    flight(id: $id) {
      id
      aircraft {
        model
        tailsign
      }
      source {
        name
        gps_code
      }
      destination {
        name
        gps_code
      }
    }
  }
`;

export default ({ match }) => {
  const { id } = match.params;
  const { loading, error, data } = useQuery(FLIGHT, { variables: { id } });

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return <pre>{JSON.stringify(data, null, 2)}</pre>;
};
