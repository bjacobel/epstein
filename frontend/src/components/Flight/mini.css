@value black, lightgrey from '../../stylesheets/colors.css';
@value margin from '../../stylesheets/measurements.css';

.flightlink {
  color: black;
  text-decoration: none;
  padding: calc(margin / 2);
}

.flightbox {
  display: flex;
  background-color: white;
  flex-direction: column;
  margin-bottom: calc(margin / 2);
  border: 1px solid lightgrey;
  max-width: 400px;
  transition: all 0.1s ease;

  &:hover {
    transform: scale(1.02);
    box-shadow: 2px 2px 1px 1px rgba(0, 0, 0, 0.1);
  }
}

.row {
  display: flex;
  justify-content: space-between;
}
