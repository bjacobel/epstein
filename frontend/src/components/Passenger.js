import React from 'react';
import { useQuery, gql } from '@apollo/client';

const PASSENGER = gql`
  query Passenger($slug: String!) {
    passenger(slug: $slug) {
      id
      name
      biography
      wikipedia_link
      image
      flights(limit: 10) {
        edges {
          id
        }
      }
    }
  }
`;

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data } = useQuery(PASSENGER, { variables: { slug } });

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return <pre>{JSON.stringify(data, null, 2)}</pre>;
};
