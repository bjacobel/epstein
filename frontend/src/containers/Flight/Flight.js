import React, { Suspense } from 'react';
import { useQuery, gql } from '@apollo/client';
import { format, parseISO } from 'date-fns';
import { Link } from 'react-router-dom';

import Details from '../../components/Details';
import Loading from '../../components/Loading';
import MetaTags from '../../components/MetaTags';
import { passengers, noslug, explainLiteral } from './style.css';
import { link } from '../../stylesheets/shared.css';

export const FLIGHT = gql`
  fragment airfield on Airfield {
    name
    iata_code
    latitude_deg
    longitude_deg
    municipality
    iso_country
  }
  query Flight($id: Int!) {
    flight(id: $id) {
      id
      distance
      date
      page
      passengers {
        edges {
          ... on VerifiedPassenger {
            name
            slug
          }
          ... on LiteralPassenger {
            literal
          }
        }
      }
      aircraft {
        id
        model
        tailsign
      }
      source {
        ...airfield
      }
      destination {
        ...airfield
      }
    }
  }
`;

const formatKm = m =>
  `${new Intl.NumberFormat().format(Number.parseFloat(m / 1000).toPrecision(4))} km`;

/* eslint-disable camelcase */
const Airfield = ({ name, municipality, iso_country }) => (
  <span>{`${name} (${municipality}, ${iso_country})`}</span>
);
/* eslint-enable camelcase */

export const describePassengers = plist => {
  const notablePassengers = plist.filter(x => x.name).map(x => x.name);
  const listablePassengers = notablePassengers.slice(0, notablePassengers.length - 1);
  const listedPassengers = listablePassengers.join(', ');
  const lastPassenger = notablePassengers[notablePassengers.length - 1];

  return `${plist.length} passengers were listed on the manifest${
    notablePassengers.length
      ? `, including ${listedPassengers}${
          listedPassengers.length ? ' and ' : ''
        }${lastPassenger}`
      : ''
  }`;
};

export default ({ match }) => {
  const { id } = match.params;
  const { loading, error, data } = useQuery(FLIGHT, { variables: { id } });

  if (loading) return <Loading text />;
  if (error) throw error;

  const isoDate = parseISO(data.flight.date);

  const SuspenseMap = React.lazy(() =>
    import(/* webpackChunkName: "map" */ '../../components/Map'),
  );

  return (
    <>
      <MetaTags
        title={`${data.flight.source.iata_code} to ${
          data.flight.destination.iata_code
        } on ${format(isoDate, 'M/d/yy')}`}
        description={describePassengers(data.flight.passengers.edges)}
        uri={`flight/${data.flight.id}`}
      />
      <Details>
        <span>date</span>
        <span>{format(isoDate, 'MMM d, y')}</span>
        <span>source</span>
        <Airfield {...data.flight.source} />
        <span>destination</span>
        <Airfield {...data.flight.destination} />
        <span>distance</span>
        <span>{formatKm(data.flight.distance)}</span>
        <span>map</span>
        <div>
          <Suspense fallback={<Loading />}>
            <SuspenseMap
              source={[data.flight.source.latitude_deg, data.flight.source.longitude_deg]}
              dest={[
                data.flight.destination.latitude_deg,
                data.flight.destination.longitude_deg,
              ]}
            />
          </Suspense>
        </div>
        <span>source</span>
        <span>
          <a
            className={link}
            href={data.flight.page}
            target="_blank"
            rel="noopener noreferrer"
          >
            original PDF document
          </a>
        </span>
        <span>passengers</span>
        <div className={passengers}>
          <ul>
            {data.flight.passengers.edges
              .filter(x => x.slug)
              .map(({ slug, name }) => (
                <li key={slug}>
                  <Link className={link} to={`/passenger/${slug}`}>
                    {name}
                  </Link>
                </li>
              ))}
            <hr />
            <span className={explainLiteral}>
              <span>As well as the following, who </span>
              <a
                className={link}
                href="/about#why-are-some-entries-not-linked-to-an-identified-passenger"
              >
                have not been linked to an identified passenger
              </a>
              :
            </span>
            {data.flight.passengers.edges
              .filter(x => !x.slug)
              .map(({ literal }) => (
                <li key={literal} className={noslug}>
                  <span>{`“${literal}”`}</span>
                </li>
              ))}
          </ul>
        </div>
        <span>aircraft</span>
        <div>
          <span>{data.flight.aircraft.model}</span>
          <br />
          <span>
            <span>tailsign </span>
            <Link className={link} to={`/aircraft/${data.flight.aircraft.id}`}>
              {data.flight.aircraft.tailsign}
            </Link>
          </span>
        </div>
      </Details>
    </>
  );
};
