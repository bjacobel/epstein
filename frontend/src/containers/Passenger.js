import React from 'react';
import { useQuery, gql } from '@apollo/client';

import Details from '../components/Details';
import PaginatedBrowser from '../components/PaginatedBrowser';
import Loading from '../components/Loading';
import MiniFlight from '../components/Mini/MiniFlight';
import { link } from '../stylesheets/shared.css';

const FLIGHT_LIMIT = 10;

const PASSENGER = gql`
  query Passenger($slug: String!, $offset: Int, $limit: Int!) {
    passenger(slug: $slug) {
      id
      name
      biography
      wikipedia_link
      image
      flightCount
      flights(offset: $offset, limit: $limit) {
        edges {
          id
        }
        pageInfo {
          count
          hasNext
        }
      }
    }
  }
`;

const range = histogram => {
  if (!histogram) return '';

  const months = Object.keys(histogram);
  const [firstDate] = months[0];
  const [lastDate] = months[months.length - 1];

  if (firstDate === lastDate) {
    return `in ${firstDate}`;
  } else {
    return `between ${firstDate} and ${lastDate}`;
  }
};

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data, fetchMore } = useQuery(PASSENGER, {
    variables: { slug, limit: FLIGHT_LIMIT },
  });

  if (loading) return <Loading text />;
  if (error) return <p>Error :(</p>;

  return (
    <Details>
      <span>name</span>
      <span>{data.passenger.name}</span>
      <span>bio</span>
      <div>
        <span>{data.passenger.biography}</span>
        <br />
        <a className={link} href={data.passenger.wikipedia_link}>
          <span>See more on Wikipedia â†’</span>
        </a>
      </div>
      <span>history</span>
      <div>
        <span>{`Appears in the flight logs at least ${data.passenger.flightCount} times`}</span>
        <span>{range(data.passenger.histogram)}</span>
      </div>
      {data.passenger.image ? (
        <>
          <span>image</span>
          <img src={data.passenger.image} alt="" />
        </>
      ) : null}
      <span>flights</span>
      <PaginatedBrowser
        browserComponent={MiniFlight}
        ids={data.passenger.flights.edges.map(x => x.id)}
        totalAvailable={data.passenger.flightCount}
        pageSize={FLIGHT_LIMIT}
        fetchMore={() =>
          fetchMore({
            variables: { offset: data.passenger.flights.edges.length + 1 },
            updateQuery: (prev, { fetchMoreResult }) => {
              if (!fetchMoreResult) return prev;
              return {
                passenger: {
                  ...prev.passenger,
                  flights: {
                    ...fetchMoreResult.passenger.flights,
                    pageInfo: {
                      ...fetchMoreResult.passenger.flights.pageInfo,
                      count:
                        prev.passenger.flights.pageInfo.count +
                        fetchMoreResult.passenger.flights.pageInfo.count,
                    },
                    edges: [
                      ...prev.passenger.flights.edges,
                      ...fetchMoreResult.passenger.flights.edges,
                    ],
                  },
                },
              };
            },
          })
        } // eslint-disable-line
      />
    </Details>
  );
};
