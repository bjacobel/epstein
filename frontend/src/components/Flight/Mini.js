import React from 'react';
import { useQuery, gql } from '@apollo/client';
import { Link } from 'react-router-dom';
import { format, parseISO } from 'date-fns';

import Loading from '../Loading';
import { flightbox, flightlink, row } from './mini.css';
import { robotoMono as mono } from '../../stylesheets/shared.css';

const FLIGHT = gql`
  query flight($id: Int!) {
    flight(id: $id) {
      id
      date
      source {
        iata_code
      }
      destination {
        iata_code
      }
      passengers {
        pageInfo {
          count
        }
      }
    }
  }
`;

export default ({ id, done }) => {
  const { loading, error, data } = useQuery(FLIGHT, {
    variables: { id },
    onCompleted: done,
  });

  if (loading) return done ? null : <Loading />;
  if (error) return <p>Error :(</p>;

  return (
    <Link to={`/flight/${data.flight.id}`} className={flightlink}>
      <div className={flightbox}>
        <div className={row}>
          <span>{format(parseISO(data.flight.date), 'MMM d, y')}</span>
          <span>{`${data.flight.passengers.pageInfo.count} passengers`}</span>
        </div>
        <div className={row}>
          <span>
            <span className={mono}>{data.flight.source.iata_code}</span>
            <span> to </span>
            <span className={mono}>{data.flight.destination.iata_code}</span>
          </span>
        </div>
      </div>
    </Link>
  );
};
