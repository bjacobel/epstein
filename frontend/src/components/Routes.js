import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import NotFound from './NotFound';
import Main from './Main';
import Flight from './Flight';
import Passenger from './Passenger';
import Analytics from '../utils/Analytics';

export class GARoute extends Route {
  render() {
    this.props.ga.pageview();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <Route {...this.props} />;
  }
}

export default class Routes extends Component {
  constructor() {
    super();
    this.ga = new Analytics();
  }

  render() {
    return (
      <>
        <BrowserRouter>
          <div>
            <Switch>
              <GARoute ga={this.ga} path="/" exact component={Main} />
              <GARoute ga={this.ga} path="/flight/:id" exact component={Flight} />
              <GARoute ga={this.ga} path="/passenger/:slug" exact component={Passenger} />
              <GARoute ga={this.ga} component={NotFound} />
            </Switch>
          </div>
        </BrowserRouter>
      </>
    );
  }
}
