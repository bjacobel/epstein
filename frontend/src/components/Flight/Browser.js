import React from 'react';

import Mini from './Mini';

export default ({ ids }) => (
  <>
    {ids.map(id => (
      <Mini id={id} key={id} />
    ))}
  </>
);
