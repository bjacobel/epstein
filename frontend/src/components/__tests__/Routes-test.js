import React from 'react';
import { shallow, mount } from 'enzyme';

import Routes, { GARoute } from 'components/Routes';
import Flight from 'components/Flight';
import Passenger from 'components/Passenger';
import Analytics from 'utils/Analytics';

jest.mock('components/Main');
jest.mock('components/Flight', () => () => null);
jest.mock('components/Passenger', () => () => null);
jest.mock('constants/wellKnownPassengers');
jest.mock('components/NotFound');
jest.mock('utils/Analytics');

const setPath = value => {
  global.history.pushState({}, value, value);
};

describe('Router', () => {
  it('matches snapshot', () => {
    expect(shallow(<Routes />)).toMatchSnapshot();
  });

  describe('Routing config', () => {
    it('has a home route', () => {
      setPath('/');
      const routes = mount(<Routes />);
      expect(routes.find('Main').length).toBe(1);
    });

    it('routes passenger/:id to the Passenger component', () => {
      setPath('/flight/125');
      const routes = mount(<Routes />);
      expect(routes.find(Flight).length).toBe(1);
      expect(routes.find(Flight).props()).toEqual(
        expect.objectContaining({
          match: expect.objectContaining({
            params: {
              id: '125',
            },
          }),
        }),
      );
    });

    it('routes flights/:id to the Flight component', () => {
      setPath('/passenger/40');
      const routes = mount(<Routes />);
      expect(routes.find(Passenger).length).toBe(1);
      expect(routes.find(Passenger).props()).toEqual(
        expect.objectContaining({
          match: expect.objectContaining({
            params: {
              id: '40',
            },
          }),
        }),
      );
    });

    it('has shorthand urls for well-known passengers', () => {
      const routes = shallow(<Routes />);
      expect(routes.find(GARoute).filter('[path^="/passenger/"]').length).toBe(4);
    });

    it('has a fallthrough 404', () => {
      setPath('/asdfasaddf');
      const routes = mount(<Routes />);
      expect(routes.find('NotFound').length).toBe(1);
    });
  });

  describe('Analytics functionality', () => {
    beforeEach(() => {
      // reset call count before each test
      Analytics.prototype.pageview.mockClear();
    });

    it('inits the analytics class', () => {
      mount(<Routes />);
      expect(Analytics).toHaveBeenCalled();
    });

    it('calls the analytics pageview fn when you navigate to a new route', () => {
      mount(<Routes />);
      setPath('/child/1');
      expect(Analytics.prototype.pageview).toHaveBeenCalledTimes(1);
    });

    it('supplies GA to routed components via props', () => {
      const routes = mount(<Routes />);
      expect(routes.find('Route').props()).toEqual(
        expect.objectContaining({
          ga: {},
        }),
      );
    });
  });
});
