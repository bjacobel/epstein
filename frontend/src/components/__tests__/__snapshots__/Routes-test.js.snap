// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router matches snapshot 1`] = `
<Fragment>
  <BrowserRouter>
    <div>
      <Switch>
        <GARoute
          component={[MockFunction]}
          exact={true}
          ga={Object {}}
          path="/"
        />
        <GARoute
          component={[Function]}
          exact={true}
          ga={Object {}}
          path="/flight/:id"
        />
        <GARoute
          component={[Function]}
          exact={true}
          ga={Object {}}
          path="/passenger/:slug"
        />
        <GARoute
          component={[MockFunction]}
          ga={Object {}}
        />
      </Switch>
    </div>
  </BrowserRouter>
</Fragment>
`;
